python/sdk/docs/tutorial.rst

```diff
@@ -0,0 +1,666 @@
+          super(Participant, self).__init__()
+          self.upper_bound = initial_upper_bound
+
+
+      def deserialize_training_input(self, data: bytes) -> Optional[np.ndarray]:
+          if not data:
+              return None
+          reader = BytesIO(data)
+          return np.load(reader, allow_pickle=False)
+
+      def serialize_training_result(self, training_result: Tuple[int, np.ndarray]) -> bytes:
+          (number_of_samples, weights) = training_result
+          writer = BytesIO()
+          writer.write(number_of_samples.to_bytes(4, byteorder="big"))
+          np.save(writer, weights, allow_pickle=False)
+          return writer.getbuffer()[:]
+
+      def train_round(self, global_weights: Optional[np.ndarray]) -> Tuple[int, np.ndarray]:
+          if global_weights is not None:
+              LOG.info("global weights: %s", global_weights)
+              self.upper_bound = global_weights[0]
+
+          value = random.uniform(0.0, self.upper_bound)
+          local_weights = np.repeat(value, 5)
+          LOG.info("local weights %s", local_weights)
+          number_of_samples = 1
+          return (number_of_samples, local_weights)
+
+  def main():
+      logging.basicConfig(
+          format="%(asctime)s.%(msecs)03d %(levelname)-8s %(message)s",
+          level=logging.DEBUG,
+          datefmt="%Y-%m-%d %H:%M:%S",
+      )
+
+      parser = argparse.ArgumentParser(description="run a participant")
+      parser.add_argument(
+          "--upper-bound",
+          required=True,
+          type=float,
+          help="Initial upper bound for picking a random float to send to the coordinator",
+      )
+      args = parser.parse_args()
+
+      participant = Participant(args.upper_bound)
+      xain_sdk.configure_logging(log_http_requests=True)
+      coordinator_url = "http://localhost:8081"
+      xain_sdk.run_participant(participant, coordinator_url)
+
+
+  if __name__ == "__main__":
+      main()
+
+
+When running with ``run-participant --upper-bound 1000``, we should
+see the global weights decreasing round after round.
+
+Going further
+-------------
+
+In this tutorial we learned how to use ``xain-sdk`` to write a participant, but that participant doesn't do real training yet. For an actual machine learning example, the the `"house pricing problem" example <https://github.com/xainag/xain-fl/tree/master/python/client_examples/keras_house_prices>`_, which uses Keras.
+
+For more details about the architecture of the platform itself, take a look at the `xainag/xain-fl Github repository <https://github.com/xainag/xain-fl>`_
+
+To see how to tune the coordinator (number of rounds, fraction of participants to select, etc.), take a look at the `sample configuration files in the xain-fl repository <https://github.com/xainag/xain-fl/tree/master/configs>`_.
+
+.. todo::
+
+   Write documentation about the aggregation process, and how one can
+   write its own aggregator
```

Comment:
perhaps put this in a ticket rather than leave in here

Reply:
Done (PB-591)